server/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -607,4 +602,53 @@ void unregisterChildNode(DiscoveryNode node) {
         }
     }
 
+    /**
+     * Start tracking a cancellable task with its tcp channel, so if the channel gets closed we can get a set of
+     * pending tasks associated that channel and cancel them as these results won't be retrieved by the parent task.
+     *
+     * @return a releasable that should be called when this pending task is completed
+     */
+    public Releasable startTrackingCancellableChannelTask(TcpChannel channel, CancellableTask task) {
+        assert cancellableTasks.containsKey(task.getId()) : "task [" + task.getId() + "] is not registered yet";
+        final ChannelPendingTaskTracker tracker = channelPendingTaskTrackers.computeIfAbsent(channel, k -> new ChannelPendingTaskTracker());
+        tracker.addTask(task);
+        if (tracker.registered.compareAndSet(false, true)) {
+            channel.addCloseListener(ActionListener.wrap(
+                r -> {
+                    final ChannelPendingTaskTracker removedTracker = channelPendingTaskTrackers.remove(channel);
+                    assert removedTracker == tracker;
+                    final Set<CancellableTask> pendingTasks = Collections.unmodifiableSet(tracker.pendingTasks);
+                    for (Consumer<Set<CancellableTask>> listener : onChannelCloseListeners) {
+                        listener.accept(pendingTasks);
+                    }
+                },
+                e -> {
+                    assert false : new AssertionError("must not be here", e);
+                }));
+        }
+        return () -> tracker.removeTask(task);
+    }
+
+    /**
+     * Register a callback which will be called when a transport channel is closed and there're some pending orphaned
+     * tasks associate with that transport channel.
+     */
+    public void registerOrphanedTasksOnChannelCloseListener(Consumer<Set<CancellableTask>> listener) {
+        onChannelCloseListeners.add(listener);
```

Comment:
Do we need to add multiple listeners ? I think  it would be cleaner to move the logic to cancel a task directly in the TaskManager and call it from `TransportCancelTasksAction` ?

Reply:
We need a single listener - I switched to SetOnce in https://github.com/elastic/elasticsearch/pull/56620/commits/1d2de32dde9b1ab7b73c483670a45b9666878cb7. I exposed this method so we can cancel child tasks.